### ISSUE TITLE
kv: reduce restarts for write-only txns

### ISSUE THREAD

**Issue by petermattis** [cockroachdb/cockroach#2859]
Consider the following. In the `./cockroach sql` shell do:

```
create database d1;
create table d1.t1 (col1 int primary key);
insert into d1.t1 values (1);
begin;
insert into d1.t1 values (2);
```

Notice the transaction has not committed. Now, in a separate shell do:

```
select * from d1.t1;
```

Finally, back in the first shell:

```
commit;
```

You should get the error `query error: kv/txn_coord_sender.go:375: transaction is already committed or aborted`. 

This is surprising. Why did a read-only operation cause the transaction to abort? Seems that the proximate cause is that the scan operation caused resolution of the write intents which pushed the transaction timestamp. @spencerkimball suggested we could allow the transaction timestamp for a write-only transaction to be pushed without restarting the transaction. 

Postgres seems to take an even finer-grained approach and tracks when a write could have had an impact on the result of a previous read from a concurrent transaction had it run first. For example:

```
create table t1 (col1 bigint primary key);
insert into t1 values (1);

begin transaction isolation level serializable;
insert into t1 (select col1+10 from t1 where col1=1);
commit;

begin transaction isolation level serializable;
insert into t1 (select col1+100 from t1 where col1=1);
commit;
```

Even if the two transactions are run concurrently, the transactions do not conflict. In Cockroach, one of the transactions will be aborted. I'm not sure if this latter case is worth worrying about, but the former case of a read-only transaction causing a write-only transaction to abort certainly seems worth addressing.

**Comment by tschottdorf** [comment]
when your `select *` comes in and sees the intent, it can only do two things: backoff and retry again later (hoping that there isn't an intent any more) or successfully push the intent. By waiting until the read goes through, you're basically guaranteeing that the writer will be pushed and that the transaction will find itself aborted. The only alternative is to be able to push the write into the future as @spencerkimball suggested (this is what would happen at snapshot isolation, but it may also apply to write-only serializable transactions).

From the KV perspective, I can't find anything surprising about your example.

Is Postgres perhaps caching your insert statement and only executing it atomically on commit? There doesn't seem to be a reason why it would execute it right away and it would certainly make sense to buffer as much as they can.

**Comment by petermattis** [comment]
You may not find this behavior surprising, but I think our user's certainly will. Seems like an artifact of our implementation, not a serializability requirement.

I don't see how Postgres could be caching the insert. It needs to be able to read the inserted data later in the transaction (read your writes).

**Comment by tschottdorf** [comment]
you were running something like

```
A begin transaction isolation level serializable;
A insert into t1 (select col1+10 from t1 where col1=1);

B begin transaction isolation level serializable;
B insert into t1 (select col1+100 from t1 where col1=1);

A commit;
B commit;
```

but I don't see a reason why Postgres wouldn't buffer your transaction and really run

```
A begin transaction isolation level serializable; insert into t1 (select col1+10 from t1 where col1=1); commit;
B begin transaction isolation level serializable; insert into t1 (select col1+100 from t1 where col1=1); commit;
```

since you're not giving it any reason to execute parts that you're writing early. If it's doing that, there's no conflict.

**Comment by petermattis** [comment]
I'm claiming that determining when the write could be "buffered" would be difficult. It is easy to force it out to the table. Add `A select * from t1 where col1=11` and `B select * from t1 where col1=101`. Still no serializability conflict.

**Comment by tschottdorf** [comment]
I think it's mostly about reads from the respective snapshots. It looks that roughly what's happening is doing snapshot serializability, but keeping track of each versioned key read. Upon commit time, if someone else already committed using that version of the key prior to a write, the commit fails. We of course can't do anything like that (not in any way performantly, anyway).

There's some substance to the idea to treat write-only transactions as snapshot serializable. The issue is that to a conflicting transaction, it's not clear whether the transaction **is** write-only (after all, it only sees the Txn entry). We could implement that on the kv level (if a Txn starts out as write-only, it can never successfully complete a read - so `CPut`, `Get`, `...` return errors) but that flag would then have to be predetermined by the `SQL` coordinator. At the end of the day, that will make it useful only for transactions which are sent wholesale. That's a significant improvement, though it won't at all help your "surprise" up there. We could, of course, introduce something like `begin write-only transaction isolation level serializable` and then it would (at least until we throw in a read just for kicks).

We could retry the Txn transparently if the first statement fails retryably as you suggested. In practice we should probably do that to avoid uncertainty-related restarts and a corresponding round-trip back to the client. For shell sessions, it won't make a huge difference, though. The reality of those is that they're going to be different from Postgres, and that the only way to run production queries is to have them in a retryable context (which the sql client could probably support if we really feel that people are going to spend a significant amount of time being surprised on the command line).

A related venue is being able to expose the `INCONSISTENT` isolation level to SQL. That could save a lot of restarts and make or break performance in various settings.

**Comment by petermattis** [comment]
I might be reading the code wrong, but it looks like we only allow pushing the timestamp of `SNAPSHOT` isolation transactions. What if we allowed pushing the timestamp of serializable transactions but then failed the transaction at commit time if a read operation had been performed?

Seems difficult for the command line SQL shell to retry transactions correctly. How do you redo the logic that the user has in their head? I must be missing something about your proposal.

**Comment by tschottdorf** [comment]
> I might be reading the code wrong, but it looks like we only allow pushing the timestamp of SNAPSHOT isolation transactions.

Yes, that's exactly the difference between `SNAPSHOT` and `SERIALIZABLE`. Checking on commit actually like a good idea and wouldn't be hard to do. We could extend the `Writing` flag to have the possible states `R|W|RW` and a serializable `W` transaction would be able to commit with a pushed timestamp.

> Seems difficult for the command line SQL shell to retry transactions correctly. How do you redo the logic that the user has in their head? I must be missing something about your proposal.

I don't think that's going to be too pretty. The user would essentially have to "program" in SQL like a KV retryable would (so no ad-hoc numbers or strings). I don't have anything in my head that I would dare call a "proposal".

**Comment by petermattis** [comment]
On Mon, Oct 19, 2015 at 4:54 PM, Tobias Schottdorf <notifications@github.com
> wrote:

> I might be reading the code wrong, but it looks like we only allow pushing
> the timestamp of SNAPSHOT isolation transactions.
>
> Yes, that's exactly the difference between SNAPSHOT and SERIALIZABLE.
> Checking on commit actually like a good idea and wouldn't be hard to do. We
> could extend the Writing flag to have the possible states R|W|RW and a
> serializable W transaction would be able to commit with a pushed
> timestamp.
>
A serializable `R` transaction will never be pushed because it never writes
an intent, correct?

**Comment by tschottdorf** [comment]
Correct.

**Comment by bdarnell** [comment]
Do we actually have any write-only transactions? Even a transaction that is write-only at the SQL level uses reads at the KV level (inserts become conditional puts)

**Comment by tschottdorf** [comment]
That's a good point, but `CPut`s in an (otherwise) write-only transaction don't ruin the above idea: Consider a transaction that's entirely `CPut`s (along with other writes ignored from now on), and assume that each of the `CPut`s is randomly pushed forward in time by concurrent readers (possibly each to its own timestamp). Then, when the transaction commits, it will commit with the maximum of these timestamps and, when resolved, the writes will all be placed at that timestamp. That is exactly the outcome of the transaction running with that timestamp in the first place in serializable mode (since no concurrent writer could have written "under" any of the intents without aborting us).

This is different from a "CPut` which reads one key `a` but writes to another key `b`. In that case, the write on `b` could get pushed, but a concurrent transaction could change the value at `a` at a timestamp lower than `b` (so it would seem that our "CPut" had read the wrong value, but still written the intent). This doesn't happen on a single-key `CPut` since the intent blocks mutation of the key for the lifetime of the transaction.

**Comment by spencerkimball** [comment]
Fixed in #13501
